File: 0542-01-matrix/0542-01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int n = mat.size();
        int m = mat[0].size();

        queue<pair<int,int>>q;
        vector<vector<int>>vis(n,vector<int>(m,0));
        vector<vector<int>>dis(n,vector<int>(m,0));

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(mat[i][j] == 0){
                    q.push({i,j});
                }
            }
        }

        while(!q.empty()){
            pair<int,int>p = q.front();
            q.pop();

            int i = p.first;
            int j = p.second;

            int del[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};

            for(int k = 0; k < 4; k++){
                int row = i + del[k][0];
                int col = j + del[k][1];

                if(row < 0 || row >= n || col < 0 || col >= m)continue;

                if(!vis[row][col] && mat[row][col]){
                    vis[row][col] = 1;
                    dis[row][col] = 1 + dis[i][j];
                    q.push({row,col});
                }else if(vis[row][col] && mat[row][col]){
                    dis[row][col] = min(dis[row][col],1+dis[i][j]);
                }
            }

        }

        return dis;
    }
};

File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    // This is bfs problem -> Min Distance
    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;

        Queue<int[]>q = new LinkedList<>();
        int[][] dis = new int[m][n];
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(mat[i][j] == 0){
                    q.add(new int[]{i,j});
                    dis[i][j] = 0;
                }
            }
        }

        int[][] del = {{1,0},{-1,0},{0,1},{0,-1}};

        while(!q.isEmpty()){
            int[] num = q.remove();
            int u = num[0];
            int v = num[1];

            for(int k = 0; k < 4; k++){
                int nrow = u + del[k][0];
                int ncol = v + del[k][1];

                if(nrow < 0 || nrow == m || ncol < 0 || ncol == n || mat[nrow][ncol] != 1 || dis[nrow][ncol] != 0)continue;

                dis[nrow][ncol] = dis[u][v] + 1;
                q.add(new int[]{nrow,ncol});
            }
        }

        return dis;
    }
}